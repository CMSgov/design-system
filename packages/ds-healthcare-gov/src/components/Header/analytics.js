import { headerSendsAnalytics } from '../flags';
import { sendLinkEvent } from '@cmsgov/design-system';

export function sendHeaderEvent(linkText, linkUrl) {
  if (headerSendsAnalytics()) {
    sendLinkEvent({
      event_name: 'header_click',
      event_type: 'ui interaction',
      category: 'consistent header',
      ga_eventAction: 'click',
      ga_eventCategory: 'consistent header',
      ga_eventLabel: linkText,
      text: linkText,
      ...(linkUrl ? { link_url: linkUrl } : {}),
    });
  }
}
