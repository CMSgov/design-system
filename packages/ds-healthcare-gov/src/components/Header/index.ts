export { default as Header } from './I18nHeader';
export { default as defaultMenuLinks } from './defaultMenuLinks';
