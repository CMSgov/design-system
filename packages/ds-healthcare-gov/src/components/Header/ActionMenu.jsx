/* eslint-disable react/no-multi-comp, jsx-a11y/no-redundant-roles */
import { Button, CloseIconThin } from '@cmsgov/design-system';
import MenuIcon from './MenuIcon';
import PropTypes from 'prop-types';
import React from 'react';
import classnames from 'classnames';
import { sendHeaderEvent } from './analytics';
import { translate } from 'react-i18next';

const menuId = 'hc-c-menu';

/**
 * ActionMenu is displayed at the top-right of the header.
 * There are two variations of this component: One for logged-out
 * users, which displays a "Log in" & "Espanol" link on desktop.
 * And another for logged-in users, which includes the user's first
 * name. The logged-in variation always displays a "Menu" toggle
 * button, and the logged-out variation only displays it on mobile.
 */
const ActionMenu = function (props) {
  return (
    <nav
      id="hc-c-header__actions"
      className="hc-c-header__actions ds-l-col ds-l-col--auto ds-u-margin-left--auto ds-u-font-weight--bold"
      role="navigation"
    >
      {props.loggedIn ? <LoggedInActionMenu {...props} /> : <LoggedOutActionMenu {...props} />}
    </nav>
  );
};

function LoggedInActionMenu(props) {
  return (
    <div>
      {props.firstName && (
        <span className="ds-u-display--none ds-u-sm-display--inline-block">{props.firstName}</span>
      )}
      <MenuButton {...props} />
    </div>
  );
}

function LoggedOutActionMenu(props) {
  if (props.links.length <= 0) return <></>;

  return (
    <div>
      <ul className="hc-c-logged-out-links ds-c-list--bare ds-u-display--none ds-u-sm-display--inline-block">
        {props.links.map((link) => {
          return (
            <li key={link.href} className="hc-c-logged-out-links__li">
              <a
                href={link.href}
                onClick={() => sendHeaderEvent(link.label, link.href)}
                className="hc-c-logged-out-links__link"
              >
                {link.label}
              </a>
            </li>
          );
        })}
      </ul>
      <MenuButton {...props} className="ds-u-display--inline-block ds-u-sm-display--none" />
    </div>
  );
}

function MenuButton({ t, open, ...props }) {
  const className = classnames(
    'hc-c-action-menu-button',
    'ds-u-sm-margin-left--2',
    props.className
  );

  function onClick(event) {
    sendHeaderEvent(open ? 'menu closed' : 'menu opened');
    props.onMenuToggleClick(event);
  }

  return (
    <Button
      aria-controls={menuId}
      aria-expanded={!!open}
      aria-label={open ? t('header.closeMenu') : t('header.openMenu')}
      className={className}
      onClick={onClick}
      size="small"
    >
      {open ? <CloseIconThin /> : <MenuIcon className="ds-u-margin-right--1" />}
      {t('header.menu')}
    </Button>
  );
}

ActionMenu.propTypes = {
  /** Applies the inverse theme styling */
  firstName: PropTypes.string,
  loggedIn: PropTypes.bool,
  onMenuToggleClick: PropTypes.func.isRequired,
  /**
   * Indicates the menu is open, which influences the label
   * and ARIA attribute of the toggle button.
   */
  open: PropTypes.bool,
  /**
   * These are the links that are visible in the upper left when there
   * is no menu button present. Currently, these only show up when the
   * user is logged out
   */
  links: PropTypes.arrayOf(
    PropTypes.shape({
      identifier: PropTypes.string,
      href: PropTypes.string.isRequired,
      label: PropTypes.node.isRequired,
      onClick: PropTypes.func,
    })
  ).isRequired,
};

LoggedInActionMenu.propTypes = {
  firstName: ActionMenu.propTypes.firstName,
  onMenuToggleClick: ActionMenu.propTypes.onMenuToggleClick,
  open: ActionMenu.propTypes.open,
};

LoggedOutActionMenu.propTypes = {
  onMenuToggleClick: ActionMenu.propTypes.onMenuToggleClick,
  open: ActionMenu.propTypes.open,
  links: PropTypes.arrayOf(
    PropTypes.shape({
      identifier: PropTypes.string,
      href: PropTypes.string.isRequired,
      label: PropTypes.node.isRequired,
      onClick: PropTypes.func,
    })
  ).isRequired,
};

MenuButton.propTypes = {
  onMenuToggleClick: ActionMenu.propTypes.onMenuToggleClick,
  open: ActionMenu.propTypes.open,
  className: PropTypes.string,
};

export default translate()(ActionMenu);
