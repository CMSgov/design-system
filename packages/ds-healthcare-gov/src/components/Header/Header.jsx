import ActionMenu from './ActionMenu';
import DeConsumerMessage from './DeConsumerMessage';
import Logo from '../Logo/Logo';
import Menu from './Menu';
import PropTypes from 'prop-types';
import React from 'react';
import { SkipNav } from '@cmsgov/design-system';
import classnames from 'classnames';
import defaultMenuLinks from './defaultMenuLinks';
import { translate } from 'react-i18next';

export const VARIATION_NAMES = {
  LOGGED_IN: 'logged-in',
  LOGGED_OUT: 'logged-out',
};

/**
 * The top-level component, responsible for maintaining the
 * header's state (like whether the mobile menu is expanded) and
 * determining which variation of the header to display
 */
export class _Header extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      openMenu: false,
    };
  }

  isLoggedIn() {
    return this.variation() === VARIATION_NAMES.LOGGED_IN;
  }

  /**
   * Content rendered within <Menu>, before the list of links
   * @returns {Node}
   */
  beforeMenuLinks() {
    if (this.isLoggedIn() && this.props.firstName) {
      return (
        <div className="ds-u-sm-display--none ds-u-border-bottom--1 ds-u-margin-x--1 ds-u-padding-y--1 hc-c-header__name">
          {this.props.firstName}
        </div>
      );
    }
  }

  /**
   * Event handler for when the "Menu" or "Close" button
   * within ActionMenu is clicked.
   */
  handleMenuToggleClick = () => {
    this.setState({ openMenu: !this.state.openMenu });
  };

  /**
   * Determines which variation of the header should be displayed,
   * based on the props being passed into the component.
   * @returns {String} Variation name
   */
  variation() {
    if (this.props.loggedIn) {
      // Logged-in state, with minimal navigation
      return VARIATION_NAMES.LOGGED_IN;
    } else {
      // Logged-out state, either Learn or Product
      return VARIATION_NAMES.LOGGED_OUT;
    }
  }

  render() {
    const classes = classnames(
      `hc-c-header hc-c-header--${this.variation()}`,
      this.props.className
    );

    const hasCustomLinks = !!this.props.links;
    const defaultLinksForVariation = defaultMenuLinks(
      this.props.initialLanguage,
      this.props.deConsumer,
      this.props.subpath,
      this.props.primaryDomain,
      this.props.switchLocaleLink,
      this.props.hideLoginLink,
      this.props.hideLogoutLink,
      this.props.hideLanguageSwitch,
      hasCustomLinks
    )[this.variation()];

    const links = hasCustomLinks
      ? this.props.links.concat(defaultLinksForVariation)
      : defaultLinksForVariation;

    return (
      <header className={classes} role="banner">
        <SkipNav href={this.props.skipNavHref} onClick={this.props.onSkipNavClick}>
          {this.props.t('header.skipNav')}
        </SkipNav>

        <div className="ds-l-container">
          <div className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2">
            <a
              href={this.props.primaryDomain ? this.props.primaryDomain : '/'}
              className="hc-c-logo-link ds-l-col ds-l-col--auto"
            >
              <Logo locale={this.props.initialLanguage} />
            </a>

            <ActionMenu
              firstName={this.props.firstName}
              onMenuToggleClick={this.handleMenuToggleClick}
              locale={this.props.initialLanguage}
              loggedIn={this.props.loggedIn}
              open={this.state.openMenu}
              links={links}
            />
          </div>
        </div>

        <Menu
          beforeLinks={this.beforeMenuLinks()}
          links={links}
          open={this.state.openMenu}
          primaryDomain={this.props.primaryDomain}
          submenuTop={this.props.submenuTop}
          submenuBottom={this.props.submenuBottom}
        />

        {this.props.deConsumer && <DeConsumerMessage deBrokerName={this.props.deBrokerName} />}
      </header>
    );
  }
}

_Header.defaultProps = {
  initialLanguage: 'en',
  skipNavHref: '#main',
};

/* eslint-disable react/no-unused-prop-types */
_Header.propTypes = {
  /**
   * Additional classes to be added to the root `<header>` element.
   */
  className: PropTypes.string,
  /**
   * The language the header will render as.
   */
  initialLanguage: PropTypes.oneOf(['en', 'es']),
  /**
   * For applications that handle their own locale switching. Overrides the
   * default locale link. The link's label is still determined by the opposite
   * of the `initialLanguage` provided, i.e. if `initialLanguage` is `en`,
   * the link's label will always be "Espa√±ol". This takes precedence over the
   * `subpath` prop.
   */
  switchLocaleLink: PropTypes.string,
  /**
   * Indicate that a user is logged-in.
   */
  loggedIn: PropTypes.bool,
  /**
   * When set to true, do not display the Login text in the upper right of the
   * header
   */
  hideLoginLink: PropTypes.bool,
  /**
   * When set to true, even if logged in the Logout link will not render
   */
  hideLogoutLink: PropTypes.bool,
  /**
   * When set to true, do not display the the switch locale link
   */
  hideLanguageSwitch: PropTypes.bool,
  /**
   * For logged-in users, pass in their first name to display in the header
   */
  firstName: PropTypes.node,
  /**
   * For applications hosted at paths other than the root `healthcare.gov`/
   * `cuidadodesalud.gov`. This string will be appended to the end of the
   * language links so as to keep the user within the same part of the site
   * when switching languages.
   */
  subpath: PropTypes.string,
  /**
   * The primary, or root domain where the majority of header links should be
   * hosted.  By default, links render with relative paths, but providing this
   * prop will force all links to render with absolute paths based on the
   * provided string. The string should include the protocol (`http://` or
   * `https://`) and the domain only, with no trailing slash. For example, if
   * the provided string is `https://test.healthcare.gov`, the login link will
   * render as `https://test.healthcare.gov/login` instead of just `/login`.
   * Note that this is only really necessary if your application is hosted on a
   * subdomain, such as `https://localhelp.healthcare.gov`, where relative links
   * would direct the user to the wrong location, e.g. the link to `/login`
   * would incorrectly direct the user to
   * `https://localhelp.healthcare.gov/login` when it should direct the user to
   * `https://healthcare.gov/login`.
   */
  primaryDomain: PropTypes.string,
  /**
   * A URL hash used for the "Skip to main content" link. This is
   * typically the `id` of your "main" content area (ie. `#main`).
   */
  skipNavHref: PropTypes.string,
  /**
   * An onClick handler used for the "Skip to main content" link. This can
   * be used in cases where one would need to manually set the focus on the
   * content area (e.g. where hash routing is being used).
   */
  onSkipNavClick: PropTypes.func,
  /**
   * Indicates when a consumer is coming from a Direct Enrollment flow.
   * This will include additional messaging and modify some of the links.
   */
  deConsumer: PropTypes.bool,
  /**
   * Used in conjunction with `deConsumer`, the Direct Enrollment broker's
   * name is used in some of the messaging displayed to the consumer.
   */
  deBrokerName: PropTypes.string,
  /**
   * Optionally pass in an array of link objects to override the default
   * set of menu links. This may be useful if you need to customize the
   * links on a page-by-page basis. To reference the default set of menu
   * links, you can import the `defaultMenuLinks` method.
   */
  links: PropTypes.arrayOf(
    PropTypes.shape({
      href: PropTypes.string.isRequired,
      label: PropTypes.node.isRequired,
      onClick: PropTypes.func,
    })
  ),
  /**
   * Optionally pass a React component to render within the menu. Useful for
   * when you need more control over what appears in the menu than what's
   * provided by the `links` prop, e.g. a search input. Will appear *above* any
   * links provided by the `defaultMenuLinks` method or the links provided by
   * the `links` prop.
   */
  submenuTop: PropTypes.node,
  /**
   * Same as `submenuTop`, except it will appear *below* any links provided by
   * the `defaultMenuLinks` method or the links provided by the `links` prop.
   */
  submenuBottom: PropTypes.node,
};

export const Header = translate()(_Header);
export default Header;
