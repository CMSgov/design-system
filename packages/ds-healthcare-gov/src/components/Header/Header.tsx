import ActionMenu from './ActionMenu';
import DeConsumerMessage from './DeConsumerMessage';
import Logo from '../Logo/Logo';
import Menu from './Menu';
import React, { useState } from 'react';
import { SkipNav, UsaBanner } from '@cmsgov/design-system';
import { t } from '../i18n';
import classnames from 'classnames';
import defaultMenuLinks from './defaultMenuLinks';

export interface Link {
  href: string;
  label: React.ReactNode;
  // TODO: I don't think this is necessary to require - PW
  ariaLabel: string;
  className?: string;
  onClick?: (...args: any[]) => any;
}

export interface HeaderProps {
  /**
   * Additional classes to be added to the root `<header>` element.
   */
  className?: string;
  /**
   * For applications that handle their own locale switching. Overrides the
   * default locale link. This takes precedence over the `subpath` prop.
   */
  switchLocaleLink?: string;
  /**
   * Indicate that a user is logged-in.
   */
  loggedIn?: boolean;
  /**
   * When set to true, do not display the Login text in the upper right of the
   * header
   */
  hideLoginLink?: boolean;
  /**
   * When set to true, even if logged in the Logout link will not render
   */
  hideLogoutLink?: boolean;
  /**
   * When set to true, do not display the the switch locale link
   */
  hideLanguageSwitch?: boolean;
  /**
   * For logged-in users, pass in their first name to display in the header
   */
  firstName?: string;
  /**
   * For applications hosted at paths other than the root `healthcare.gov`/
   * `cuidadodesalud.gov`. This string will be appended to the end of the
   * language links so as to keep the user within the same part of the site
   * when switching languages.
   */
  subpath?: string;
  /**
   * The primary, or root domain where the majority of header links should be
   * hosted.  By default, links render with relative paths, but providing this
   * prop will force all links to render with absolute paths based on the
   * provided string. The string should include the protocol (`http://` or
   * `https://`) and the domain only, with no trailing slash. For example, if
   * the provided string is `https://test.healthcare.gov`, the login link will
   * render as `https://test.healthcare.gov/login` instead of just `/login`.
   * Note that this is only really necessary if your application is hosted on a
   * subdomain, such as `https://localhelp.healthcare.gov`, where relative links
   * would direct the user to the wrong location, e.g. the link to `/login`
   * would incorrectly direct the user to
   * `https://localhelp.healthcare.gov/login` when it should direct the user to
   * `https://healthcare.gov/login`.
   */
  primaryDomain?: string;
  /**
   * A URL hash used for the "Skip to main content" link. This is
   * typically the `id` of your "main" content area (ie. `#main`).
   */
  skipNavHref?: string;
  /**
   * An onClick handler used for the "Skip to main content" link. This can
   * be used in cases where one would need to manually set the focus on the
   * content area (e.g. where hash routing is being used).
   */
  onSkipNavClick?: (...args: any[]) => any;
  /**
   * Indicates when a consumer is coming from a Direct Enrollment flow.
   * This will include additional messaging and modify some of the links.
   */
  deConsumer?: boolean;
  /**
   * Used in conjunction with `deConsumer`, the Direct Enrollment broker's
   * name is used in some of the messaging displayed to the consumer.
   */
  deBrokerName?: string;
  /**
   * Optionally pass in an array of link objects to override the default
   * set of menu links. This may be useful if you need to customize the
   * links on a page-by-page basis. To reference the default set of menu
   * links, you can import the `defaultMenuLinks` method.
   */
  links?: Link[];
  /**
   * Optionally pass a React component to render within the menu. Useful for
   * when you need more control over what appears in the menu than what's
   * provided by the `links` prop, e.g. a search input. Will appear *above* any
   * links provided by the `defaultMenuLinks` method or the links provided by
   * the `links` prop.
   */
  submenuTop?: React.ReactNode;
  /**
   * Same as `submenuTop`, except it will appear *below* any links provided by
   * the `defaultMenuLinks` method or the links provided by the `links` prop.
   */
  submenuBottom?: React.ReactNode;
  /**
   * Element added to display content on Header bottom section
   */
  headerBottom?: React.ReactNode;
  /**
   * Open and handler function for fully controlled menu behavior
   */
  isMenuOpen?: boolean;
  onMenuToggle?: () => void;
  /**
   * Additional classes to be added to the Logo component
   */
  logoClassName?: string;
  /**
   * Temporary feature flag for showing or not showing the USA Banner above the
   * header. Defaults to true
   */
  showUsaBanner?: boolean;
}

export const VARIATION_NAMES = {
  LOGGED_IN: 'logged-in',
  LOGGED_OUT: 'logged-out',
};

/**
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/header/healthcare-header/?theme=healthcare).
 */
export const Header = (props: HeaderProps) => {
  const [openMenu, setOpenMenu] = useState(false);
  const isControlledMenu = props.isMenuOpen !== undefined && props.onMenuToggle !== undefined;

  /**
   * Determines which variation of the header should be displayed,
   * based on the props being passed into the component.
   * @returns {String} Variation name
   */
  function variation(): string {
    if (props.loggedIn) {
      // Logged-in state, with minimal navigation
      return VARIATION_NAMES.LOGGED_IN;
    } else {
      // Logged-out state, either Learn or Product
      return VARIATION_NAMES.LOGGED_OUT;
    }
  }

  function isLoggedIn() {
    return variation() === VARIATION_NAMES.LOGGED_IN;
  }

  /**
   * Content rendered within <Menu>, before the list of links
   * @returns {Node}
   */
  function beforeMenuLinks(): JSX.Element {
    if (isLoggedIn() && props.firstName) {
      return (
        <div className="ds-u-sm-display--none ds-u-border-bottom--1 ds-u-margin-x--1 ds-u-padding-y--1 hc-c-header__name">
          {props.firstName}
        </div>
      );
    }
  }

  /**
   * Event handler for when the "Menu" or "Close" button
   * within ActionMenu is clicked.
   */
  function handleMenuToggleClick() {
    if (!isControlledMenu) {
      setOpenMenu(!openMenu);
    } else {
      props.onMenuToggle();
    }
  }

  const classes = classnames(`hc-c-header hc-c-header--${variation()}`, props.className);

  const hasCustomLinks = !!props.links;
  const defaultLinksForVariation = defaultMenuLinks({
    deConsumer: props.deConsumer,
    subpath: props.subpath,
    primaryDomain: props.primaryDomain,
    switchLocaleLink: props.switchLocaleLink,
    hideLoginLink: props.hideLoginLink,
    hideLogoutLink: props.hideLogoutLink,
    hideLanguageSwitch: props.hideLanguageSwitch,
    customLinksPassedIn: hasCustomLinks,
  })[variation()];

  const links = hasCustomLinks
    ? props.links.concat(defaultLinksForVariation)
    : defaultLinksForVariation;

  return (
    <>
      {props.showUsaBanner && <UsaBanner id="hc-c-header__usa-banner" />}
      <header className={classes} role="banner" aria-label="global">
        <SkipNav href={props.skipNavHref} onClick={props.onSkipNavClick}>
          {t('header.skipNav')}
        </SkipNav>

        <div className="ds-l-container">
          <div className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2">
            <a
              href={props.primaryDomain ? props.primaryDomain : '/'}
              className="hc-c-logo-link ds-l-col ds-l-col--auto"
            >
              <Logo className={props.logoClassName ?? ''} />
            </a>

            <nav
              aria-label="Profile, applications, and coverage"
              id="hc-c-header__actions"
              className="hc-c-header__actions ds-l-col ds-l-col--auto ds-u-margin-left--auto ds-u-font-weight--bold"
            >
              <ActionMenu
                t={t}
                firstName={props.firstName}
                onMenuToggleClick={handleMenuToggleClick}
                loggedIn={props.loggedIn}
                open={isControlledMenu ? props.isMenuOpen : openMenu}
                links={links}
              />
              <Menu
                beforeLinks={beforeMenuLinks()}
                links={links}
                open={isControlledMenu ? props.isMenuOpen : openMenu}
                submenuTop={props.submenuTop}
                submenuBottom={props.submenuBottom}
              />
            </nav>
          </div>
        </div>

        {props.deConsumer && <DeConsumerMessage t={t} deBrokerName={props.deBrokerName} />}
        {props.headerBottom}
      </header>
    </>
  );
};

Header.defaultProps = {
  showUsaBanner: true,
  skipNavHref: '#main',
};

export default Header;
