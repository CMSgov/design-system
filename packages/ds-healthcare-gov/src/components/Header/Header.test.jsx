import { _Header as Header } from './Header';
import React from 'react';
import { SkipNav } from '@cmsgov/design-system';
import { shallow } from 'enzyme';

jest.mock('../i18n', () => ({
  translate: (text) => text,
}));

function render(props) {
  props = Object.assign(
    {
      t: (key) => key,
    },
    props
  );

  return shallow(<Header {...props} />);
}

describe('Header', function () {
  it('renders full/homepage header', () => {
    expect(render()).toMatchSnapshot();
  });

  it('renders Direct Enrollment banner', () => {
    expect(
      render({
        deConsumer: true,
        deBrokerName: 'Foo',
      })
    ).toMatchSnapshot();
  });

  it('renders logged-in header with firstName', () => {
    expect(render({ loggedIn: true, firstName: 'John' })).toMatchSnapshot();
  });

  it('renders logged-in header without firstName', () => {
    expect(render({ loggedIn: true })).toMatchSnapshot();
  });

  it('renders Spanish header', () => {
    expect(render({ initialLanguage: 'es' })).toMatchSnapshot();
  });

  it('toggles openMenu state when handleMenuToggleClick is called', () => {
    const wrapper = render();

    expect(wrapper.state().openMenu).toBe(false);

    wrapper.instance().handleMenuToggleClick();

    expect(wrapper.state().openMenu).toBe(true);
  });

  it('passes correct props to SkipNav', () => {
    const props = {
      skipNavHref: 'javascript:void(0)',
      onSkipNavClick: () => {},
    };
    const wrapper = render(props);
    const skipNav = wrapper.find(SkipNav);
    expect(skipNav.length).toEqual(1);
    expect(skipNav.props()).toMatchObject({
      href: props.skipNavHref,
      onClick: props.onSkipNavClick,
    });
  });

  it('re-renders with updated links', () => {
    const props = { loggedIn: true };
    const wrapper = render(props);

    let menu = wrapper.find('Menu');
    expect(menu.prop('links').length).toBe(4);

    wrapper.setProps({ links: [{ href: '/foo', label: 'Foo' }] });
    menu = wrapper.find('Menu');

    // You always get the logout link and locale link with any custom links unless you
    // explicitly disable them
    expect(menu.prop('links').length).toBe(3);
  });

  it('should add spanish toggle if logged in', () => {
    const props = { loggedIn: true };
    const wrapper = render(props);

    const menu = wrapper.find('Menu');
    expect(menu.prop('links').length).toBe(4);

    expect(menu.prop('links')[2].label).toEqual('header.español');
  });

  it('should not add Spanish toggle when hideLanguageSwitch set', () => {
    const props = { hideLanguageSwitch: true };
    const wrapper = render(props);

    const menu = wrapper.find('Menu');
    expect(menu.prop('links').length).toBe(1);

    expect(menu.prop('links')[0].label).not.toEqual('header.español');
  });

  it('should not add Login Link when hideLoginLink set', () => {
    const props = { hideLoginLink: true };
    const wrapper = render(props);

    const menu = wrapper.find('Menu');
    expect(menu.prop('links').length).toBe(1);

    expect(menu.prop('links')[0].label).not.toEqual('header.login');
  });

  it('should not add Logout Link when hideLogoutLink set', () => {
    const props = { loggedIn: true, hideLogoutLink: true, links: [] };
    const wrapper = render(props);

    const menu = wrapper.find('Menu');
    expect(menu.prop('links').length).toBe(1);

    expect(menu.prop('links')[0].label).not.toEqual('header.logout');
  });

  it('should have "logout" as last item when logged in', () => {
    const props = { loggedIn: true };
    const wrapper = render(props);

    const menu = wrapper.find('Menu');
    const menuLinks = menu.dive().find('MenuLinks');
    const lastLink = menuLinks.dive().find('a').last();

    expect(lastLink).toBeDefined();
    expect(lastLink.text()).toEqual('header.logout');
  });

  it('renders links with absolute URLs if provided a primaryDomain prop', () => {
    expect(render({ primaryDomain: 'https://www.healthcare.gov' })).toMatchSnapshot();
  });
});
