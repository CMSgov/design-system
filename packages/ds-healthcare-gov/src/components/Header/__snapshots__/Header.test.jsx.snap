// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders Direct Enrollment banner 1`] = `
<header
  className="hc-c-header hc-c-header--logged-out"
  role="banner"
>
  <SkipNav
    href="#main"
  >
    header.skipNav
  </SkipNav>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2"
    >
      <a
        className="hc-c-logo-link ds-l-col ds-l-col--auto"
        href="/"
      >
        <Logo
          locale="en"
        />
      </a>
      <Component
        links={
          Array [
            Object {
              "href": "https://www.cuidadodesalud.gov/es/",
              "label": "header.español",
            },
            Object {
              "href": "/login?check_de=1",
              "identifier": "login",
              "label": "header.login",
            },
          ]
        }
        locale="en"
        onMenuToggleClick={[Function]}
        open={false}
      />
    </div>
  </div>
  <Menu
    links={
      Array [
        Object {
          "href": "https://www.cuidadodesalud.gov/es/",
          "label": "header.español",
        },
        Object {
          "href": "/login?check_de=1",
          "identifier": "login",
          "label": "header.login",
        },
      ]
    }
    open={false}
  />
  <Component
    deBrokerName="Foo"
  />
</header>
`;

exports[`Header renders Spanish header 1`] = `
<header
  className="hc-c-header hc-c-header--logged-out"
  role="banner"
>
  <SkipNav
    href="#main"
  >
    header.skipNav
  </SkipNav>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2"
    >
      <a
        className="hc-c-logo-link ds-l-col ds-l-col--auto"
        href="/"
      >
        <Logo
          locale="es"
        />
      </a>
      <Component
        links={
          Array [
            Object {
              "href": "https://www.healthcare.gov/",
              "label": "header.english",
            },
            Object {
              "href": "/login",
              "identifier": "login",
              "label": "header.login",
            },
          ]
        }
        locale="es"
        onMenuToggleClick={[Function]}
        open={false}
      />
    </div>
  </div>
  <Menu
    links={
      Array [
        Object {
          "href": "https://www.healthcare.gov/",
          "label": "header.english",
        },
        Object {
          "href": "/login",
          "identifier": "login",
          "label": "header.login",
        },
      ]
    }
    open={false}
  />
</header>
`;

exports[`Header renders full/homepage header 1`] = `
<header
  className="hc-c-header hc-c-header--logged-out"
  role="banner"
>
  <SkipNav
    href="#main"
  >
    header.skipNav
  </SkipNav>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2"
    >
      <a
        className="hc-c-logo-link ds-l-col ds-l-col--auto"
        href="/"
      >
        <Logo
          locale="en"
        />
      </a>
      <Component
        links={
          Array [
            Object {
              "href": "https://www.cuidadodesalud.gov/es/",
              "label": "header.español",
            },
            Object {
              "href": "/login",
              "identifier": "login",
              "label": "header.login",
            },
          ]
        }
        locale="en"
        onMenuToggleClick={[Function]}
        open={false}
      />
    </div>
  </div>
  <Menu
    links={
      Array [
        Object {
          "href": "https://www.cuidadodesalud.gov/es/",
          "label": "header.español",
        },
        Object {
          "href": "/login",
          "identifier": "login",
          "label": "header.login",
        },
      ]
    }
    open={false}
  />
</header>
`;

exports[`Header renders links with absolute URLs if provided a primaryDomain prop 1`] = `
<header
  className="hc-c-header hc-c-header--logged-out"
  role="banner"
>
  <SkipNav
    href="#main"
  >
    header.skipNav
  </SkipNav>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2"
    >
      <a
        className="hc-c-logo-link ds-l-col ds-l-col--auto"
        href="https://www.healthcare.gov"
      >
        <Logo
          locale="en"
        />
      </a>
      <Component
        links={
          Array [
            Object {
              "href": "https://www.cuidadodesalud.gov/es/",
              "label": "header.español",
            },
            Object {
              "href": "https://www.healthcare.gov/login",
              "identifier": "login",
              "label": "header.login",
            },
          ]
        }
        locale="en"
        onMenuToggleClick={[Function]}
        open={false}
      />
    </div>
  </div>
  <Menu
    links={
      Array [
        Object {
          "href": "https://www.cuidadodesalud.gov/es/",
          "label": "header.español",
        },
        Object {
          "href": "https://www.healthcare.gov/login",
          "identifier": "login",
          "label": "header.login",
        },
      ]
    }
    open={false}
    primaryDomain="https://www.healthcare.gov"
  />
</header>
`;

exports[`Header renders logged-in header with firstName 1`] = `
<header
  className="hc-c-header hc-c-header--logged-in"
  role="banner"
>
  <SkipNav
    href="#main"
  >
    header.skipNav
  </SkipNav>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2"
    >
      <a
        className="hc-c-logo-link ds-l-col ds-l-col--auto"
        href="/"
      >
        <Logo
          locale="en"
        />
      </a>
      <Component
        firstName="John"
        links={
          Array [
            Object {
              "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
              "label": "header.myApplicationsAndCoverage",
            },
            Object {
              "href": "/marketplace/auth/global/en_US/myProfile#settings",
              "label": "header.myProfile",
            },
            Object {
              "href": "https://www.cuidadodesalud.gov/es/",
              "label": "header.español",
            },
            Object {
              "href": "/logout",
              "identifier": "logout",
              "label": "header.logout",
            },
          ]
        }
        locale="en"
        loggedIn={true}
        onMenuToggleClick={[Function]}
        open={false}
      />
    </div>
  </div>
  <Menu
    beforeLinks={
      <div
        className="ds-u-sm-display--none ds-u-border-bottom--1 ds-u-margin-x--1 ds-u-padding-y--1 hc-c-header__name"
      >
        John
      </div>
    }
    links={
      Array [
        Object {
          "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
          "label": "header.myApplicationsAndCoverage",
        },
        Object {
          "href": "/marketplace/auth/global/en_US/myProfile#settings",
          "label": "header.myProfile",
        },
        Object {
          "href": "https://www.cuidadodesalud.gov/es/",
          "label": "header.español",
        },
        Object {
          "href": "/logout",
          "identifier": "logout",
          "label": "header.logout",
        },
      ]
    }
    open={false}
  />
</header>
`;

exports[`Header renders logged-in header without firstName 1`] = `
<header
  className="hc-c-header hc-c-header--logged-in"
  role="banner"
>
  <SkipNav
    href="#main"
  >
    header.skipNav
  </SkipNav>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row ds-u-align-items--center ds-u-flex-wrap--nowrap ds-u-padding-y--2"
    >
      <a
        className="hc-c-logo-link ds-l-col ds-l-col--auto"
        href="/"
      >
        <Logo
          locale="en"
        />
      </a>
      <Component
        links={
          Array [
            Object {
              "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
              "label": "header.myApplicationsAndCoverage",
            },
            Object {
              "href": "/marketplace/auth/global/en_US/myProfile#settings",
              "label": "header.myProfile",
            },
            Object {
              "href": "https://www.cuidadodesalud.gov/es/",
              "label": "header.español",
            },
            Object {
              "href": "/logout",
              "identifier": "logout",
              "label": "header.logout",
            },
          ]
        }
        locale="en"
        loggedIn={true}
        onMenuToggleClick={[Function]}
        open={false}
      />
    </div>
  </div>
  <Menu
    links={
      Array [
        Object {
          "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
          "label": "header.myApplicationsAndCoverage",
        },
        Object {
          "href": "/marketplace/auth/global/en_US/myProfile#settings",
          "label": "header.myProfile",
        },
        Object {
          "href": "https://www.cuidadodesalud.gov/es/",
          "label": "header.español",
        },
        Object {
          "href": "/logout",
          "identifier": "logout",
          "label": "header.logout",
        },
      ]
    }
    open={false}
  />
</header>
`;
